_posts/2022-10-01-arrow-parquet-encoding-part-2.md

```diff
@@ -0,0 +1,344 @@
+
+```text
+Field(name: "a", nullable: true, datatype: List(
+  Field(name: "element", nullable: true, datatype: Int32),
+)
+```
+
+Documents of this format could be stored in this Parquet schema
+
+```text
+message schema {
+  optional group a (LIST) {
+    repeated group list {
+      optional int32 element;
+    }
+  }
+}
+```
+
+As before, Arrow chooses to represent this in a hierarchical fashion with a list of monotonically increasing integers called *offsets* in a `ListArray`, and stores all the values that appear in the lists in a single child array. Each consecutive pair of elements in this offset array identifies a slice of the child array for that array index
+
+For example, the list of offsets `[0, 2, 3, 3]` contains 3 pairs of offsets, `(0,2)`, `(2,3)`, and `(3,3)`, and is therefore a ListArray of length 3 with the following values:
+
+```text
+0: [child[0], child[1]]
+1: []
+2: [child[2]]
+```
+
+For the example above with 4 JSON documents, this would be encoded in Arrow as
+
+
+```text
+┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─
+                          ┌──────────────────┐ │
+│    ┌─────┐   ┌─────┐    │ ┌─────┐   ┌─────┐│
+     │  1  │   │  0  │    │ │  1  │   │  1  ││ │
+│    ├─────┤   ├─────┤    │ ├─────┤   ├─────┤│
+     │  0  │   │  1  │    │ │  0  │   │ ??  ││ │
+│    ├─────┤   ├─────┤    │ ├─────┤   ├─────┤│
+     │  1  │   │  1  │    │ │  1  │   │  2  ││ │
+│    ├─────┤   ├─────┤    │ └─────┘   └─────┘│
+     │  1  │   │  1  │    │ Validity   Values│ │
+│    └─────┘   ├─────┤    │                  │
+               │  3  │    │ child[0]         │ │
+│    Validity  └─────┘    │ PrimitiveArray   │
+                          │                  │ │
+│              Offsets    └──────────────────┘
+     "a"                                       │
+│    ListArray
+ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
+```
+
+More technical detail is available in the [ListArray format specification](https://arrow.apache.org/docs/format/Columnar.html#variable-size-list-layout).
+
+
+### Repetition Levels
+
+In order to encode lists, Parquet stores an integer *repetition level* in addition to a definition level. A repetition level identifies where in the hierarchy of repeated fields the current value is to be inserted. A value of `0` would imply a new list in the top-most repeated field, a value of 1 a new element within the top-most repeated field, a value of 2 a new element within the second top-most repeated field, and so on.
+
+Each repeated field also has a corresponding definition level, however, in this case rather than indicating a null value, they indicate an empty array.
+
+
+
+```text
+┌─────────────────────────────────────┐
+│  ┌─────┐      ┌─────┐               │
+│  │  3  │      │  0  │               │
+│  ├─────┤      ├─────┤               │
+│  │  0  │      │  0  │               │
+│  ├─────┤      ├─────┤      ┌─────┐  │
+│  │  1  │      │  1  │      │  1  │  │
```

Comment:
```suggestion
│  │  1  │      │  0  │      │  1  │  │
```
A sanity check you can do is that a repetition level of 0 indicates the start of a new row, therefore the number of zeros must match the number of rows